### ISSUE TITLE
MongoError: failed to connect to server [localhost:27017] on first connect

### ISSUE THREAD

**Issue by kholisrag** [nosqlclient/nosqlclient#354]
I install docker Mongoclient 2.0.0, I run it with :

> docker run -d -p 3000:3000 -v /var/lib/mongodb:/data/db -e MONGO_URL=mongodb://localhost:27017/ mongoclient/mongoclient

but it's force close, I check the docker logs, there is the logs:

> MongoError: failed to connect to server [localhost:27017] on first connect
    at Object.Future.wait (/opt/meteor/dist/bundle/programs/server/node_modules/fibers/future.js:449:15)
    at new MongoConnection (packages/mongo/mongo_driver.js:211:27)
    at new MongoInternals.RemoteCollectionDriver (packages/mongo/remote_collection_driver.js:4:16)
    at Object.<anonymous> (packages/mongo/remote_collection_driver.js:38:10)
    at Object.defaultRemoteCollectionDriver (packages/underscore.js:784:19)
    at new Mongo.Collection (packages/mongo/collection.js:99:40)
    at meteorInstall.lib.imports.collections.settings.js (lib/imports/collections/settings.js:6:25)
    at fileEvaluate (packages/modules-runtime.js:197:9)
    at Module.require (packages/modules-runtime.js:120:16)
    at Module.Mp.importSync (/opt/meteor/dist/bundle/programs/server/npm/node_modules/meteor/modules/node_modules/reify/lib/runtime.js:82:24)
    - - - - -
    at [object Object].<anonymous> (/opt/meteor/dist/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/topologies/server.js:326:35)
    at emitOne (events.js:77:13)
    at [object Object].emit (events.js:169:7)
    at [object Object].<anonymous> (/opt/meteor/dist/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/pool.js:270:12)
    at [object Object].g (events.js:260:16)
    at emitTwo (events.js:87:13)
    at [object Object].emit (events.js:172:7)
    at Socket.<anonymous> (/opt/meteor/dist/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/connection.js:175:49)
    at Socket.g (events.js:260:16)
    at emitOne (events.js:77:13)

Anyone can help?

**Comment by rsercano** [comment]
Hi there @kholisrag 

what do you mean it forces close ? Does docker stop ? Yesterday I've done an update in the code base, that can be the related. 

Except from that, this error means it's not possible to connect your connection parameters, probably there should be an error on your configuration (hostname, port etc...).

**Comment by kholisrag** [comment]
@rsercano the docker can't run...
must run in root privilege?

anyway, I just run the web version... It's work...

**Comment by rsercano** [comment]
@kholisrag 

it actually doesn't need root privileges, but you need to bind a data volume that has to be created by Mongoclient docker container.

**Comment by rsercano** [comment]
I assume that can be closed @kholisrag please feel free to re-open it if you get this issue again.

**Comment by suewonjp** [comment]
I got the same error.

The docker container starts with no problem when I set up with no data binding, but it won't start when I specify volume like @petrukngantuk did.

And I can't find VOLUME keyword in the [Dockerfile](https://hub.docker.com/r/mongoclient/mongoclient/dockerfile)...

Is that OK?

Thanks

**Comment by rsercano** [comment]
@suewonjp dockerfile needs to be updated you're right @suewonjp but I'm already working with volumes and having no trouble, would you please share what exact error do you get ? 

In fact, if you have a chance to raise another issue with those concerns I would be more than happy ! This issue is a bit old.

**Comment by satyadeep** [comment]
> I got the same error.
> 
> The docker container starts with no problem when I set up with no data binding, but it won't start when I specify volume like @petrukngantuk did.
> 
> And I can't find VOLUME keyword in the [Dockerfile](https://hub.docker.com/r/mongoclient/mongoclient/dockerfile)...
> 
> Is that OK?
> 
> Thanks

Hi @suewonjp Could you please mention the command you used / lines you used in the docker-compose file, to start up the container and also the error you received when you tried to start the container with bind mounts?

**Comment by suewonjp** [comment]
Hi @satyadeep 

Well, I tried it again this morning, and it worked!

There must have been something wrong with my command. (maybe, incorrect host folder path for the bind mount...)

Sorry for the hassle...

Thanks

**Comment by suewonjp** [comment]
Hi.

Well, it did work on my local machine, but it didn't on my remote machine.

Command:

```
docker run --name nosqlclient -d -p 3000:3000 -v ~/work/volumes/nosqlclient:/data/db mongoclient/mongoclient
```

Logs:

```
{"error":{"name":"MongoNetworkError","errorLabels":["TransientTransactionError"]},"level":"error","message":"uncaughtException: failed to connect to server [127.0.0.1:27017] on first connect [MongoNetworkError: connect ECONNREFUSED 127.0.0.1:27017]\nError: failed to connect to server [127.0.0.1:27017] on first connect [MongoNetworkError: connect ECONNREFUSED 127.0.0.1:27017]\n    at Pool.<anonymous> (/opt/meteor/dist/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/topologies/server.js:564:11)\n    at emitOne (events.js:115:13)\n    at Pool.emit (events.js:210:7)\n    at Connection.<anonymous> (/opt/meteor/dist/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/pool.js:317:12)\n    at Object.onceWrapper (events.js:318:30)\n    at emitTwo (events.js:125:13)\n    at Connection.emit (events.js:213:7)\n    at Socket.<anonymous> (/opt/meteor/dist/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/connection.js:246:50)\n    at Object.onceWrapper (events.js:316:30)\n    at emitOne (events.js:115:13)\n    at Socket.emit (events.js:210:7)\n    at emitErrorNT (internal/streams/destroy.js:64:8)\n    at _combinedTickCallback (internal/process/next_tick.js:138:11)\n    at process._tickDomainCallback (internal/process/next_tick.js:218:9)","stack":"Error: failed to connect to server [127.0.0.1:27017] on first connect [MongoNetworkError: connect ECONNREFUSED 127.0.0.1:27017]\n    at Pool.<anonymous> (/opt/meteor/dist/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/topologies/server.js:564:11)\n    at emitOne (events.js:115:13)\n    at Pool.emit (events.js:210:7)\n    at Connection.<anonymous> (/opt/meteor/dist/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/pool.js:317:12)\n    at Object.onceWrapper (events.js:318:30)\n    at emitTwo (events.js:125:13)\n    at Connection.emit (events.js:213:7)\n    at Socket.<anonymous> (/opt/meteor/dist/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/connection.js:246:50)\n    at Object.onceWrapper (events.js:316:30)\n    at emitOne (events.js:115:13)\n    at Socket.emit (events.js:210:7)\n    at emitErrorNT (internal/streams/destroy.js:64:8)\n    at _combinedTickCallback (internal/process/next_tick.js:138:11)\n    at process._tickDomainCallback (internal/process/next_tick.js:218:9)","exception":true,"date":"Mon Mar 25 2019 01:43:29 GMT+0000 (UTC)","process":{"pid":1,"uid":1000,"gid":999,"cwd":"/opt/meteor/dist/bundle/programs/server","execPath":"/opt/nodejs/bin/node","version":"v8.4.0","argv":["/opt/nodejs/bin/node","/opt/meteor/dist/bundle/main.js","program.json"],"memoryUsage":{"rss":114196480,"heapTotal":83267584,"heapUsed":59407544,"external":21311531}},"os":{"loadavg":[0,0,0],"uptime":451183},"trace":[{"column":11,"file":"/opt/meteor/dist/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/topologies/server.js","function":null,"line":564,"method":null,"native":false},{"column":13,"file":"events.js","function":"emitOne","line":115,"method":null,"native":false},{"column":7,"file":"events.js","function":"Pool.emit","line":210,"method":"emit","native":false},{"column":12,"file":"/opt/meteor/dist/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/pool.js","function":null,"line":317,"method":null,"native":false},{"column":30,"file":"events.js","function":"Object.onceWrapper","line":318,"method":"onceWrapper","native":false},{"column":13,"file":"events.js","function":"emitTwo","line":125,"method":null,"native":false},{"column":7,"file":"events.js","function":"Connection.emit","line":213,"method":"emit","native":false},{"column":50,"file":"/opt/meteor/dist/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/connection.js","function":null,"line":246,"method":null,"native":false},{"column":30,"file":"events.js","function":"Object.onceWrapper","line":316,"method":"onceWrapper","native":false},{"column":13,"file":"events.js","function":"emitOne","line":115,"method":null,"native":false},{"column":7,"file":"events.js","function":"Socket.emit","line":210,"method":"emit","native":false},{"column":8,"file":"internal/streams/destroy.js","function":"emitErrorNT","line":64,"method":null,"native":false},{"column":11,"file":"internal/process/next_tick.js","function":"_combinedTickCallback","line":138,"method":null,"native":false},{"column":9,"file":"internal/process/next_tick.js","function":"process._tickDomainCallback","line":218,"method":"_tickDomainCallback","native":false}]}
```

Still, as I described it before it just works without -v option, but it failed with -v option.

( I verified the host folder for the mount does exist... )

The deference between my local machine and remote machine is docker version as well as OS.

Local : Docker version 18.09.2, build 6247962  (macOS Mojave)

Remote : Docker version 18.06.1-ce, build e68fc7a215d7133c34aa18e3b72b4a21fd0c6136 (Amazon Linux)

I'm managing many docker containers (like Jenkins, mongodb, nginx, etc) with volumes on my remote machine, but no issue like this.

And I'm still wonder... is that really OK without VOLUME keyword in your dockerfile?

I've been believing all along it is necessary when you intend to use bind mounts for you container ... Correct me if I'm wrong  :-)

**Comment by satyadeep** [comment]
> And I'm still wonder... is that really OK without VOLUME keyword in your dockerfile?
> 
> I've been believing all along it is necessary when you intend to use bind mounts for you container ... Correct me if I'm wrong :-)

@suewonjp It is adviced to have a VOLUME directive in the Dockerfile when using it for database images but no, It is not a necessity to have the VOLUME directive in the dockerfile. This is especially true in the case of bind mounts.

Having a volume directive like below, in the Dockerfile, will create a persistent volume with the contents of /data/db.
`VOLUME [ "/data/db" ]
`
i.e When the above directive is used, a volume is created on the host, within the docker root path (/var/lib/docker by default), and then mounted to the path specified in the Dockerfile (/data/db). This volume will persist even after the container is removed.

However, this VOLUME directive in the Dockerfile will NOT have any impact on bind mounts as the host folder (~/work/volumes/nosqlclient, in your case) is mounted at the path (/data/db). Having a VOLUME directive in the Dockerfile will only have an impact if you are running the container like this and would like the mongodb data to persist

`docker run --name nosqlclient -d -p 3000:3000 mongoclient/mongoclient`

Coming to the error, the only way I'm able to reproduce the error is, if the host folder path is wrong or non-existent, but I've seen that you've already checked that. Could you just recheck, just in case something is missing?

**Comment by rsercano** [comment]
Thank you @satyadeep

**Comment by suewonjp** [comment]
@satyadeep Thank you for great explanation.

> Could you just recheck, just in case something is missing?

I rechecked a few times. No luck. 

In fact, some error message was missing from what I've posted before.

```
MongoDB: exception in initAndListen: 20 Attempted to create a lock file on a read-only directory: /data/db, terminating
```

However, the host folder is clearly not readonly. And other docker containers are working with other host folders. 

Anyway, looks like the issue is not reproducible to you. I'll dig into it further.
